hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
             recordBuilder.status(ResponseCodeEnum.UNKNOWN);
         }
 
-        // TODO: handle long scheduled transactions
+        // store all records at once
+        final var recordListResult = recordListBuilder.build();
+        recordManager.endUserTransaction(recordListResult.recordStream());
+
+        // Verify if the transaction is a duplicate and add it to the cache
+        if (preHandleResult != null) { // this should always be true, checking just in case
+            try {
+                checkDuplicatesAndIncludeInCache(
+                        preHandleResult,
+                        recordListResult.mainRecord().recordStreamItem().record(),
+                        consensusNow);
+            } catch (final PreCheckException e) {
+                recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder);
+            }
+        }
```

Comment:
The duplicate check is in the wrong place. It **must** happen before we execute the transaction, and before we create the record stream. The purpose of the duplicate check is so that if the same transaction is sent twice, we only execute it once. Otherwise, we will have a double spend problem, which would be very bad. Also, it is a waste of resources to execute a transaction twice. Also, we must know the status of the transaction before we send it to the record manager.

This all needs to move up around line 155 or so. I think what we want to do is:
0. Check whether this transaction is a node due diligence failure. In that case, charge the node and skip to finalization.
1. Check whether this transaction is a duplicate, and add it to the cache.
2. Whether or not it is a duplicate, verify the payer signature.
3. If the payer signature is invalid or the payer has insufficient funds, charge the node.
4. If the payer signature is valid and the payer has sufficient funds, update the record build to indicate this is a duplicate transaction, and charge the user accordingly (finalize transaction, submit record file, etc)
5. If it was not a duplicate, then after verifying the payer signature, execute the transaction, finalize it, etc.

Reply:
Oh! I made a mistake above. Actually, we need to break the duplication check into TWO independent steps. Oof.

It is true that we need to check for duplicates BEFORE we do "handle". The list of steps above are right about that. But, we SHOULD NOT add the transaction to the cache until AFTER we know the final status (about where you have it on line 216).

Reply:
Okay, I pushed some changes. I split the cache addition from the duplication check. Now if the transaction is duplicated and signature verification is okay I'm setting the status of the recordBuilder to DUPLICATE_TRANSACTION and I'm calling the `createPenaltyPayment` method and leaving the flow to continue to process the penalty.

When we implement the createPenaltyPayment method(https://github.com/hashgraph/hedera-services/issues/6811) maybe we should have an argument to decide whether to charge the node or the user(or maybe we can have a second method for that).

About point **0** - this is already implemented in getCurrentPreHandleResult method
About point **3** - I'm not sure if this is something I should implement here or if this is for a different story?